asserts/model.go

```diff
@@ -25,16 +25,310 @@ import (
+		defaultChannel = "stable"
+	}
+
+	if defaultChannel != "" {
+		_, err := channel.Parse(defaultChannel, "-")
+		if err != nil {
+			return nil, fmt.Errorf("invalid default channel for snap %q: %v", name, err)
+		}
+	} else {
+		trackCh, err := channel.ParseVerbatim(track, "-")
+		if err != nil || trackCh.Track == "" || trackCh.Risk != "" || trackCh.Branch != "" {
+			return nil, fmt.Errorf("invalid locked track for snap %q: %s", name, track)
+		}
+	}
+
+	presence, err := checkOptionalStringWhat(snap, "presence", what)
+	if err != nil {
+		return nil, err
+	}
+	if presence != "" && !strutil.ListContains(validSnapPresences, presence) {
+		return nil, fmt.Errorf("presence of snap %q must be one of required|optional", name)
+	}
+
+	return &ModelSnap{
+		Name:           name,
+		SnapID:         snapID,
+		SnapType:       typ,
+		Modes:          modes, // can be empty
+		DefaultChannel: defaultChannel,
+		Track:          track,
+		Presence:       presence, // can be empty
+	}, nil
+}
+
+// unextended case support
+
+func checkSnapWithTrack(headers map[string]interface{}, which string) (*ModelSnap, error) {
+	_, ok := headers[which]
+	if !ok {
+		return nil, nil
+	}
+	value, ok := headers[which].(string)
+	if !ok {
+		return nil, fmt.Errorf(`%q header must be a string`, which)
+	}
+	l := strings.SplitN(value, "=", 2)
+
+	name := l[0]
+	track := ""
+	if err := validateSnapName(l[0], which); err != nil {
+		return nil, err
+	}
+	if len(l) > 1 {
+		track = l[1]
+		if strings.Count(track, "/") != 0 {
+			return nil, fmt.Errorf(`%q channel selector must be a track name only`, which)
+		}
+		channelRisks := []string{"stable", "candidate", "beta", "edge"}
+		if strutil.ListContains(channelRisks, track) {
+			return nil, fmt.Errorf(`%q channel selector must be a track name`, which)
+		}
+	}
+
+	defaultChannel := ""
+	if track == "" {
+		defaultChannel = "stable"
+	}
+
+	return &ModelSnap{
+		Name:           name,
+		SnapType:       which,
+		Modes:          defaultModes,
+		DefaultChannel: defaultChannel,
+		Track:          track,
+		Presence:       "required",
+	}, nil
+}
+
+func validateSnapName(name string, headerName string) error {
+	if err := naming.ValidateSnap(name); err != nil {
+		return fmt.Errorf("invalid snap name in %q header: %s", headerName, name)
+	}
+	return nil
+}
+
+func checkRequiredSnap(name string, headerName string, snapType string) (*ModelSnap, error) {
+	if err := validateSnapName(name, headerName); err != nil {
+		return nil, err
+	}
+
+	return &ModelSnap{
+		Name:           name,
+		SnapType:       snapType,
+		Modes:          defaultModes,
+		DefaultChannel: "stable",
+		Presence:       "required",
+	}, nil
+}
+
 // Model holds a model assertion, which is a statement by a brand
 // about the properties of a device model.
 type Model struct {
 	assertionBase
-	classic          bool
-	requiredSnaps    []string
+	classic bool
+
+	baseSnap   *ModelSnap
+	gadgetSnap *ModelSnap
+	kernelSnap *ModelSnap
+
+	allSnaps         []*ModelSnap
```

Comment:
Reading this it isn't clear why one is a list of concrete types while the other of snap reference interfaces. Perhaps a comment would help?